ImportXml.py

```diff
@@ -0,0 +1,182 @@
+                cat_lab = uniCodeRe(cat.contingentQuestion.find("text").get_text())
+            cates[cat_lab] = cat_val
+        return cates
+
+    def findSubquestions(que) -> dict:
+        res_sub = dict()
+        subs = que.find_all("subQuestion")
+        for sub in subs:
+            res_sub[sub["varName"]] = sub.find("text").get_text()
+        return res_sub
+
+    def concatQuesitonText(ques) -> str:
+        res_desc = ""
+        for que in ques:
+            res_desc += uniCodeRe(BeautifulSoup(que.string, "html.parser").get_text())
+        return res_desc
+
+    Ans = dict()
+    try:
+        f = open(filepath, "r")
+    except OSError:
+        print(f"{filepath} is an invalid xml file")
+        sys.exit()
+
+    soup = BeautifulSoup(f, "xml")
+    secs = soup.find_all("section")
+    res_sections = list()
+    res_questions = list()
+    for sec in secs:
+        sec_id = sec.attrs["id"]
+        sec_title = sec.sectionInfo.contents[3].string
+        for info in sec.find_all("sectionInfo"):
+            sec_text = info.find_all("text")
+            text_p = ""
+            for t in sec_text:
+                text_p += textParse(t.string)
+            sec_info = text_p
+        res_sections.append({"id": sec_id, "title": sec_title, "info": sec_info})
+        ques = sec.find_all("question")
+        for que in ques:
+            que_cat = {}
+            que_sub = {}
+            resps = que.find_all("response")
+            que_que = uniCodeRe(
+                BeautifulSoup(que.find("text").string, "html.parser").get_text()
+            )
+
+            if que.find("directive") is not None:
+                que_descs = que.find("directive").find_all("text")
+                que_desc = concatQuesitonText(que_descs)
+            else:
+                que_desc = ""
+
+            for resp in resps:
+                que_names = {
+                    True: resps[0].attrs["varName"] + " multi answers",
+                    False: resp.attrs["varName"],
+                }
+                que_name = que_names[len(resps) != 1]
+
+                # if under question next level exist "format" means
+                # it is a "longtext" question else "category" question
+                if resp.category is None:
+                    que_type = resp.format.string  # longtext
+                    que_cat = {}
+                    que_sub = {}
+                else:
+                    que_type = "catogory"
+                    que_cat.update(findCategories(resp))
+                    que_sub.update(findSubquestions(que))
+
+            res_questions.append(
+                {
+                    "name": que_name,
+                    "type": que_type,
+                    "question": que_que,
+                    "description": que_desc,
+                    "categories": que_cat,
+                    "sub_questions": que_sub,
+                    "section_id": sec_id,
+                }
+            )
```

Comment:
This part I would also isolate to a separate function. Something like:

```python
for question in questions:
    question_structure = _get_question_structure(question)
    question_structure.update(section_id=section_id)
    res_questions.append(question_structure)
```

Reply:
ok